### ISSUE TITLE
MPI: Fatal error in PMPI_Testall

### ISSUE THREAD

**Issue by obscureed** [firemodels/fds#9035]
I get an MPI error with some input files, for example:
[southeast0002.fds.txt](https://github.com/firemodels/fds/files/5644559/southeast0002.fds.txt)

````
&HEAD CHID='southeast0002', TITLE='southeast0002'  /

&TIME T_END=2.0, /

&MESH ID='mesh03' IJK=24,21,25,    XB=-2.400,0.000,   -2.000,0.100,   0.000,2.500,    MPI_PROCESS=0   /
&MESH ID='mesh05' IJK=36,40,25,    XB=-6.000,-2.400,   -0.300,3.700,   0.000,2.500,    MPI_PROCESS=1   /
&MESH ID='mesh06' IJK=63,14,25,    XB=-8.700,-2.400,   -1.700,-0.300,   0.000,2.500,    MPI_PROCESS=2   /
&MESH ID='mesh07' IJK=27,49,25,    XB=-8.700,-6.000,   -0.300,4.600,   0.000,2.500,    MPI_PROCESS=3   /
&MESH ID='mesh10' IJK=26,38,15,    XB=0.000,2.600,   0.600,4.400,   1.000,2.500,    MPI_PROCESS=4   /
&MESH ID='mesh09' IJK=28,26,25,    XB=0.000,2.800,   -2.000,0.600,   0.000,2.500,    MPI_PROCESS=4   /

&ZONE XYZ=-1.000,-1.000,0.400    /

&TAIL
````
If I run this in serial, it's fine. If I run it in parallel, for example on Windows in a cmdFDS session with
`fds_local -p 5 southeast.fds`
or on Ubuntu with
`mpiexec -n 5 fds southeast.fds`
then I get an error (same for both OSs):
```
Starting FDS ...

 MPI Process      0 started on engcfd4
 MPI Process      3 started on engcfd4
 MPI Process      4 started on engcfd4
 MPI Process      2 started on engcfd4
 MPI Process      1 started on engcfd4

 Reading FDS input file ...

Abort(17) on node 4 (rank 4 in comm 0): Fatal error in PMPI_Testall: See the MPI_ERROR field in MPI_Status for the error code
```

The example I've given is minimal, in the sense that it runs OK in parallel if I delete the final &MESH.
Both installations have been working fine in parallel for other cases. Both are recent: FDS6.7.5-0-g71f025606 or FDS6.7.5-0-g71f025606-release.
I've tried the three fixes discussed here: https://github.com/firemodels/fds/issues/5853. Maybe the second fix there needs to be updated, because I get a message `[0] MPI startup(): I_MPI_SHM_LMT environment variable is not supported.`, but I don't know where to start in updating it.

The example runs OK in parallel with FDS6.7.4 on Ubuntu (ah, OK, with the fixes still in place and the same `not supported` message), and so does the real case it came from. That's my current workaround.

Any ideas, please?

**Comment by rmcdermo** [comment]
Thanks for the report.  Looking into it now.

**Comment by rmcdermo** [comment]
@obscureed The case runs with latest source on our Linux cluster.  It also runs with latest source on macOS in parallel using a hostfile with 5 slots.

However, if I dial back to 71f025606 and try to run on macOS it hangs.  So, could you try an experiment on your end?  Please install the latest test bundle, which is provided in the link at the top of the [FDS Release Notes](https://github.com/firemodels/fds/wiki/FDS-Release-Notes) wiki, and see if that works for you.  Thanks

**Comment by mcgratta** [comment]
The case runs for me with the latest source code.

>

**Comment by obscureed** [comment]
Thanks for your efforts! Yes, the example works in parallel with FDS 6.7.5-559 on Linux. The full case runs as well (though I shouldn't use unvalidated versions for paid work). At first I thought it was slow, then I remembered to `export OMP_NUM_THREADS=1`.

I've just built an almost identical case, and that runs in the existing release version. Do you have any idea what is causing the difference?

**Comment by mcgratta** [comment]
We (the developers) typically run test cases with the same number of MPI
processes as meshes. That is the fastest way. So we sometimes find bugs for
cases where the number of processes does not match the number of meshes. I
fix a bug like this every few months, but I cannot off the top of my head
recall the exact details of the last fix.

>

**Comment by obscureed** [comment]
I agree that one-process-per-mesh is the fastest way if your case can be partitioned into roughly equal meshes without increasing the cell count. But how about when the geometry does not permit this?

A frequent case for me is a long corridor, with an apartment leading off it, with the corridor connected to a tall stairwell. At the ground level, a short corridor leads from the stairwell to the exterior. That's three orthogonal strands. The bounding box could be 95% empty. I use larger cells for some regions -- for example, the ground-floor corridor. In the "natural" partition (that is, choosing meshes that mostly contain fluid), this ground-floor corridor is less than 1% of the mesh count.

So my choices are:
(1) have mesh blocks with very different cell counts, and combine multiple meshes into a single MPI process;
(2) apply one process per mesh, where some meshes have 20 times as many cells as others;
(3) find mesh blocks with more equal cell counts, but accept that some of these will be >80% obstacle and the total cell count will be higher.
Currently I go for choice (1). In the current case, the natural partition is simple enough that choice (2) is easily possible -- would you recommend it? Sometimes, choice (2) would not be so simple, but the number of meshes can be cut down by pushing towards choice (3). This would take some user time and ingenuity, but so does choice (1).

If we accept that choice (1) is sometimes sensible (or even sometimes unavoidable without 80%-obstacle meshes), then perhaps it should be reflected in your range of test cases?

I've strayed away from the bug report -- sorry! But in case you're interested, I've attached my current mesh partition: 15 meshes combined into 8 MPI processes. This is not an extreme case at all. I didn't apply much time to balancing the cell counts across processes. I'd be very interested to hear your recommended approach (for run-time, solution stability and user effort). Thanks for your time!
[meshes_only.fds.txt](https://github.com/firemodels/fds/files/5652432/meshes_only.fds.txt)

**Comment by mcgratta** [comment]
There is nothing wrong with Option 1, and we have a few verification cases to test that it works. But, there are so many possible mesh configurations that it is difficult to comprehensively test them all. That is why we occasionally find bugs. For our validation cases, we generally just divide up all regions of the domain into meshes with comparable number of cells. In cases where we cannot do that, we just assign the MPI process to each and acknowledge that the MPI process assigned to the small mesh will not work 100% of the time. That is still faster, assuming you have enough cores, than doubling up meshes on a single core.

Does your meshes_only.fds case fail? If not, I suggest that you run the case for a certain amount of simulated time with some simple flow field and just test the different configurations.

**Comment by obscureed** [comment]
I've tried running a case similar to this in three ways:
(a) 8 MPI processes (from 15 meshes) on an 8-process machine;
(b) 8 MPI processes (from 15 meshes) on a 16-process machine;
(c) 15 MPI processes on a 16-process machine.
Runs (b) and (c) ran at identical speeds, give or take, and 40% faster than run (a). My interpretation is that the 16-process machine (8 physical cores) is doing a good job of redirecting spare compute power to the harder-working processes. (I've tried using two OpenMP threads per process, as a variant of (b), and it was slower -- but, as you said, all cases are different.) If speed is the important question, then in this case choices (1) and (2) do equally well. If value-for-money is important (and a 16-process machine costs about twice an 8-process machine), then run (a) using choice (1) is sensible.

The meshes_only.fds case fails (in FDS6.7.5, in parallel) -- including if I repartition with one process per mesh. I can see that chasing a bug into the MPI libraries could be more trouble than it's worth, so if there are no obvious fixes or workarounds (other than "try a different version") I'm OK with closing this issue. Thanks for your help.
